mtp_api/apps/prison/forms.py

```diff
@@ -1,6 +1,82 @@
+import csv
+import decimal
+import io
+
 from django import forms
+from django.contrib.admin.widgets import AdminFileWidget
+from django.db import transaction
+from django.forms.models import ModelChoiceField
 from django.utils.translation import gettext_lazy as _
 
+from prison.models import Prison, PrisonerBalance, validate_prisoner_number
+
 
 class LoadOffendersForm(forms.Form):
     modified_only = forms.BooleanField(label=_('Only load updated offender locations'), required=False)
+
+
+class PrisonerBalanceUploadForm(forms.Form):
+    csv_file = forms.FileField(label=_('Prisoner balances file'), allow_empty_file=True, widget=AdminFileWidget)
+    prison = ModelChoiceField(queryset=Prison.objects.filter(use_nomis_for_balances=False))
+
+    def clean_csv_file(self):
+        csv_file = self.cleaned_data.get('csv_file')
+        if csv_file:
+            try:
+                return self.parse_balances(csv_file)
+            except Exception as e:
+                self.add_error('csv_file', f'Could not parse file: {e}')
+        return None
+
+    @classmethod
+    def parse_balances(cls, csv_file):
+        csv_contents = csv_file.read().decode()
+        csv_contents = io.StringIO(csv_contents)
+        csv_reader = csv.DictReader(csv_contents)
+        balances = []
+        for line in csv_reader:
+            prisoner_number = line['prisonernumber'].strip().upper()
+            # check that prisoner number follows expected pattern (else indicates malformed file)
+            validate_prisoner_number(prisoner_number)
+            amount = line['totalamount']
+            amount_pence = decimal.Decimal(amount.strip()) * 100
+            # check that amount_pence is not negative (assumed impossible)
+            # and has no decimal places (indicates malformed file)
+            numerator, denomiator = amount_pence.as_integer_ratio()
+            if numerator <= 0:
+                raise ValueError(f'Negative balance {amount}')
+            if denomiator != 1:
+                raise ValueError(f'Cannot turn {amount} into pence')
+            balances.append({
+                'prisoner_number': prisoner_number,
+                'amount': int(amount_pence),
+            })
+        return balances
+
+    @transaction.atomic
+    def save(self):
+        prison = self.cleaned_data['prison']
+        balances = self.cleaned_data['csv_file']
+        balances = [
+            PrisonerBalance(
+                prison=prison,
+                prisoner_number=balance['prisoner_number'],
+                amount=balance['amount'],
+            )
+            for balance in balances
+        ]
+        # delete all balances at this prison (because small balances will not be provided in spreadsheet)
+        delete_result = PrisonerBalance.objects.filter(prison=prison).delete()
```

Comment:
Any reason why we couldn't combine these two delete queries into one?

Reply:
I thought about that, but couldn't see how (might be lack of django knowledge). Can we filter and get results from both queries (prison OR prisoner_number__in) then might make sense.

Reply:
no real reason other than it's probs a bit clearer as 2 calls. otherwise we'd have to do something like `.filter(Q(prison=prison) | Q(prisoner_number__in=set(balance.prisoner_number for balance in balances)))`… and i dunno if postgres would use the indices nicely